contracts/exchangeIssuance/ExchangeIssuanceZeroEx.sol

```diff
@@ -0,0 +1,561 @@
+        allowedIssuanceModules[_issuanceModule].isDebtIssuanceModule = _isDebtIssuanceModule;
+    }
+
+    /**
+     * Removes an issuance module from the whitelist
+     *
+     * @param _issuanceModuleAddress    Address of issuance module to remove
+     */
+    function _removeIssuanceModule(address _issuanceModuleAddress) internal {
+        allowedIssuanceModules[_issuanceModuleAddress].isAllowed = false;
+    }
+
+    /**
+     * Sets a max approval limit for an ERC20 token, provided the current allowance
+     * is less than the required allownce.
+     *
+     * @param _token    Token to approve
+     * @param _spender  Spender address to approve
+     */
+    function _safeApprove(IERC20 _token, address _spender, uint256 _requiredAllowance) internal {
+        uint256 allowance = _token.allowance(address(this), _spender);
+        if (allowance < _requiredAllowance) {
+            _token.safeIncreaseAllowance(_spender, type(uint96).max - allowance);
+        }
+    }
+
+    /**
+     * Issues an exact amount of SetTokens using WETH.
+     * Acquires SetToken components at the best price accross uniswap and sushiswap.
+     * Uses the acquired components to issue the SetTokens.
+     *
+     * @param _setToken          Address of the SetToken being issued
+     * @param _amountSetToken    Amount of SetTokens to be issued
+     *
+     */
+    function _buyComponentsForInputToken(ISetToken _setToken, uint256 _amountSetToken, bytes[] memory _quotes, IERC20 _inputToken, address _issuanceModule) internal returns (uint256 totalInputTokenSold) {
+        uint256 componentAmountBought;
+        uint256 inputTokenAmountSold;
+
+        (address[] memory components, uint256[] memory componentUnits) = getRequiredIssuanceComponents(_issuanceModule, _setToken, _amountSetToken);
+
+        uint256 inputTokenBalanceBefore = _inputToken.balanceOf(address(this));
+        for (uint256 i = 0; i < components.length; i++) {
+            address component = components[i];
+            uint256 units = componentUnits[i];
+
+            // If the component is equal to the input token we don't have to trade
+            if(component == address(_inputToken)) {
+                totalInputTokenSold = totalInputTokenSold.add(units);
+                componentAmountBought = units;
+            }
+            else {
+                uint256 componentBalanceBefore = IERC20(component).balanceOf(address(this));
+                _fillQuote(_quotes[i]);
+                uint256 componentBalanceAfter = IERC20(component).balanceOf(address(this));
+                componentAmountBought = componentBalanceAfter.sub(componentBalanceBefore);
+                require(componentAmountBought >= units, "ExchangeIssuance: UNDERBOUGHT COMPONENT");
+            }
+        }
+        uint256 inputTokenBalanceAfter = _inputToken.balanceOf(address(this));
+        totalInputTokenSold = totalInputTokenSold.add(inputTokenBalanceBefore.sub(inputTokenBalanceAfter));
```

Comment:
Not certain about this but might it be cheaper to return the value directly rather than declare and set `totalInputTokenSold`?

Reply:
This variable is used to be able to account for the case when one of the components is also the input / output token. (see `if(component == address(_inputToken))...` above. In this case the change in input token balance alone would not account for the total const of the issuance. 

Hope that makes sense. If not let me know and I will rephrase.

Reply:
See:
https://github.com/FlattestWhite/index-coop-smart-contracts/blob/exchange-issuance-zero-ex/contracts/exchangeIssuance/ExchangeIssuanceZeroEx.sol#L413

Reply:
Ah yes sorry! lgtm.